core/src/main/scala/kafka/server/AbstractAsyncFetcher.scala

```diff
@@ -31,19 +36,50 @@ import org.apache.kafka.common.requests.EpochEndOffset._
 import org.apache.kafka.common.requests.{EpochEndOffset, FetchRequest, FetchResponse, OffsetsForLeaderEpochRequest}
 import org.apache.kafka.common.{InvalidRecordException, TopicPartition}
 
-import java.nio.ByteBuffer
-import java.util.Optional
-import java.util.concurrent.TimeUnit
-import java.util.function.Consumer
 import scala.collection.JavaConverters._
 import scala.collection.{Map, Seq, Set, mutable}
 import scala.math.min
 
 
+sealed trait FetcherEvent extends Comparable[FetcherEvent] {
+  def priority: Int // an event with a higher priority value is more important
+  def state: FetcherState
+  override def compareTo(other: FetcherEvent): Int = {
+    // an event with a higher proirity value should be dequeued first in a PriorityQueue
+    other.priority - this.priority
```

Comment:
From the PriorityQueue doc:
> If multiple elements are tied for least value, the head is one of those elements -- ties are broken arbitrarily. 

Does it make sense to break ties with arrival order?  Perhaps a sequence number generated from an atomic integer?

Reply:
For now, there is no need to break ties since at most one high-priority request can be in this queue.

That is,  the AbstractFetcherManager has a lock field to ensure at most one request handler thread can be performing one of the 3 operations at any given time: adding partitions, removing partitions, or getting partition counts.

Reply:
What about all the other requests?  As it stands they will not be executed in any deterministic order.

Reply:
So far there is only one low-priority task, which is the TruncateAndFetch. Going forward, I don't expect dependencies between different low-priority fetcher events.

Reply:
@smccauliff after an offline sync up, I've decided to add the sequence number to differentiate events with the same priority. Please take another look at the latest code.